src/AutopinPlus/ControlStrategy.cpp

```diff
@@ -34,65 +34,128 @@
 #include <QString>
 #include <QStringList>
 #include <QFileInfo>
+#include <QSet>
+#include <QTextStream>
 
 namespace AutopinPlus {
 
+ControlStrategy::Pinning ControlStrategy::pinning;
+QMutex ControlStrategy::mutex;
+
 ControlStrategy::ControlStrategy(const Configuration &config, const ObservedProcess &proc, OS::OSServices &service,
 								 const PerformanceMonitor::monitor_list &monitors, AutopinContext &context)
 	: config(config), proc(proc), service(service), monitors(monitors), context(context), name("ControlStrategy") {}
 
-void ControlStrategy::init() {
-	// Setup pinning history
-	createPinningHistory();
+const ControlStrategy::Task ControlStrategy::emptyTask = {0, 0};
+
+bool ControlStrategy::Task::operator==(const ControlStrategy::Task &rhs) const {
+	return (this->tid == rhs.tid) && (this->pid == rhs.pid);
 }
 
-QString ControlStrategy::getName() { return name; }
+bool ControlStrategy::Task::operator!=(const ControlStrategy::Task &rhs) const {
+	return (this->tid != rhs.tid) || (this->pid != rhs.pid);
+}
 
-void ControlStrategy::slot_TaskCreated(int) {}
+bool ControlStrategy::Task::isCpuFree() const { return (*this == emptyTask); }
 
-void ControlStrategy::slot_TaskTerminated(int) {}
+void ControlStrategy::init() {
+	if (pinning.size() == 0) {
+		context.info("Reading /proc/cpu");
```

Comment:
https://www.gnu.org/software/libc/manual/html_node/Processor-Resources.html
This may also be relevant and a more straight forward solution if you only need the number of CPU cores.